### DISCUSSION TITLE
Get formatted citation with lua function

### DISCUSSION THREAD

**Discussion by estedeahora** [jgm/pandoc#9749] [Ideas]
I'm trying to generate xml output that is compatible with the [Crossref xml schema](https://data.crossref.org/reports/help/schema_doc/5.3.1/common5_3_1_xsd.html). For this I generated a custom template, which I use together with the jats-xml writer and some filters to fix the most important differences (actually I could use the html writer, since in xml-Crossref you don't have to put the body text, but only metadata, but it works fine with jats so I leave it). 

For the `<citation>` tag, Crossref has a fairly customised way of ordering the fields (see [recommendations](https://www.crossref.org/documentation/register-maintain-records/maintaining-your-metadata/add-references/)). With a lua filter I adapted the 'references' field to match the tags that crossref uses. One of the tags that uses this scheme is `<unstructured_citation>` which is basically the formatted citation. I wrote a lua filter that generates the references in the body (using `pandoc.utils.citeproc`) and then generates a table with the form `id = formatted reference`. With this table I add, for each reference in 'meta', the field manually. 

This works, but I feel it's a bit far-fetched.

I should make it clear that, although this seems a rather particular use case, a similar situation occurs for the schema of [Scielo 1.9] (https://scielo.readthedocs.io/projects/scielo-publishing-schema/pt-br/latest/) and earlier, for which I also plan to generate the xml). In this case, it is mandatory to include the tag `<mixed-citation>` inside each `<ref>` element.

The proposal is to have a function inside the lua api that allows to generate references for each reference element directly (without having to generate a div with the references in the blocks). I imagine that something like this: `pandoc.utils.formatedcitation(meta.references[1])`, could give me the formatted citation.

**Comment by jgm** [comment]
I'm not sure I understand.  But if you just want the contents of the Div that normally encloses the whole citation, you can easily extract that with the current Lua API.

**Comment by estedeahora** [comment]
Thank you @jgm for the reply. Sorry if I didn't make myself clear. I slightly edited my question because I realised that by not using \` to point to the xml tags inside `<` they were not visible.

 I was thinking and they are actually two different use cases. I will describe both.

The **first use case** is Crossref. For that I have this little filter:

```{lua}
local stringify = pandoc.utils.stringify
local RawBlock = pandoc.RawBlock
local citeproc = pandoc.utils.citeproc

local format_citation = {}
local citation = ''

function Pandoc(doc)

    local refs = doc.meta.references
    
    if not refs then
        return doc
    end

    -- make a table with formatted citations
    local doc2 = citeproc(doc)

    for _, b in ipairs(doc2.blocks) do
        if b.identifier == "refs" then 
            format_citation = {}
            for k, cite in ipairs(b.content) do
                format_citation[cite.identifier] = stringify(cite)
            end
        end
    end
    
    -- Convert references into crossref citation format
    local refs2 = {}

    for k, ref in ipairs(refs) do
        local id = stringify(ref.id)

        citation = '<citation key="ref=' .. id .. '">\n'

        -- some code with others fields --

        if format_citation['ref-' .. id] then
            citation = citation .. '<unstructured_citation>' .. format_citation['ref-' .. id] .. '</unstructured_citation>\n'
        end
        citation = citation .. '</citation>'
        
        refs2[k] = RawBlock('jats', citation)      

    end
    doc.meta.refs2 = refs2

    return doc
end
```

Then in the template I use: 

```
<citation_list>
$for(refs2)$
$it$
$endfor$
</citation_list>
```

This works for this use case. My *question* is whether I am making this more complicated than it should be.

The **second use case** is (I think) a bit more complex. In this case it is a specification of the jats DTD 1.1 made by Scielo. In this specification it is required that each `<ref>` tag includes a `<mixed-citation>` subtag. The problem here is that when we use `jats_publishing+element_citations` Pandoc automatically generates the `<ref-list>` structure (with the `<ref>` inside) and places it inside `back`. However, these references do not include the `<mixed-citation>` tag. I understand that Pandoc's behaviour is logical and makes sense (since the elements of the reference are present), but Scielo considers it invalid. So, I would need to know how to access this back and add the `<mixed-citation>` in a similar way as I did in the previous case. So, as you point out, this is not the same problem as above.

The solution I see for now would be to intervene the resulting xml outside of Pandoc, but I like my workflow to go through Pandoc.

**Reply by jgm** [reply]
Pandoc does include a mixed-citation tag, doesn't it?
https://github.com/jgm/pandoc/blob/main/src/Text/Pandoc/Writers/JATS.hs#L289-L292
Maybe this is being blocked because you're using `+element_citations` -- try without that?

**Reply by estedeahora** [reply]
> Maybe this is being blocked because you're using `+element_citations` -- try without that?

Ok, thank you. I hadn't seen that. 

However, when I remove 'element_citations', it also removes the `<element-citation>` tag (and all its contents). The problem is that this tag is also required by Scielo. Scielo requires both `<element-citation>` and `<mixed-citation>` to be present. 

I tried to access `back` from a Lua filter so I could change this manually, but no matter if citeproc is before, this element is not available.

(Clearly this question is no longer the original one. Apologies for not being clear from the beginning.)

**Reply by fredguth** [reply]
Have you solved the Scielo problem?  I need to export a markdown as a Scielo 1.9 valid xml (based on JATS 1.1).  I export to JATS and either get `<mixed-citation>` or `<elements-citation>` (jats+element_citations extension).

**Comment by tarleb** [comment]
Commenting mainly to ask if you would update us when you make progress.

The way we're currently handling this in JOSS is similar to what you describe, i.e., we're using the `<unstructured_citation>` element. Our defaults file for that transformation is here: https://github.com/openjournals/inara/blob/main/data/defaults/crossref.yaml
Note that some of that code is a bit dated, e.g., it predates `pandoc.utils.citeproc`. I think that, if I was to write this today, I'd use a custom writer instead of a filter.

**Reply by estedeahora** [reply]
Hi @tarleb. I'm passing you the link to the [lua filter](https://github.com/estedeahora/guri/blob/main/inst/files-pkg/filters/crossref-citation.lua) where I developed this. I think you might be interested in the project as I see that it has some points of contact with the one you are developing. I would be interested in exchanging on the subject if you are interested. I still couldn't figure out how to write pandoc writers.
